content/post/2019-05-28.Rmd

```diff
@@ -0,0 +1,322 @@
+---
+title: "Simple yet elegant Object-Oriented programming in R with S3"
+authors: ["Iñaki Úcar"]
+date: 2019-05-28
+categories: ["R"]
+tags: ["R", "S3", "object-oriented programming"]
+---
+
+```{r, setup, include = FALSE}
+knitr::opts_chunk$set(echo = TRUE)
+```
+
+The R language is peculiar in many ways, and its approach to object-oriented (OO) programming is just one of them. Indeed, base R supports not one, but three different OO systems: S3, S4 and RC classes. And yet, probably none of them would qualify as a fully-fledged OO system before the astonished gaze of an expert in languages such as Python, C++ or Java. In this tutorial, we will review the S3 system, the simplest yet most elegant of them. The use case of the [quantities framework](https://github.com/r-quantities) (CRAN packages `units`, `errors` and `quantities`) will serve as the basis to study the main concepts behind S3 programming in R: classes, generics, methods and inheritance.
+
+## Introduction
+
+You probably heard of John Chambers' slogans about R, i.e., *everything that exists is an object* and *everything that happens is a function call* (and a function is an object too). In most languages, *object* is an OO concept referring to an instance of a certain class that encapsulates its whole lifetime: construction, associated data, behaviour and destruction. In R, instead, roughly speaking, an object is just a base type with possibly some attributes. In other words, everything is about the data in R, while construction, behaviour and destruction are defined elsewhere.
+
+Let us start with the common example: what are *factors* in R? What makes them so special?
+
+```{r}
+x <- factor(c("apple", "orange", "orange", "pear"))
+typeof(x)
+attributes(x)
+unclass(x)
+```
+
+As you can see, factors are just integer vectors with some metadata stored as an attribute named `levels`. Note also that there is also a special attribute called `class` which is central to defining how these objects behave. Where and how to do it is what S3 programming is about. This decoupling makes S3 less formal and more error-prone, but on the other hand, it is much more flexible, because it allows you e.g. to extend classes in other packages without requiring them to modify a single bit.
```

Comment:
Después del primer punto, repites 'Also' dos veces muy consecutivas. Creo que sin el 2do (o sin el primero) de ellos se entiende lo que quieres decir.

Reply:
- Quitado un "also".
- "is central in defining", ni pa ti ni pa mí. :)